### PR TITLE
add support for indexing SAR data into elasticsearch 1.7.x

### PR DESCRIPTION
**Pull request by arcolife**
w.r.t.: #227 
This PR adds SAR data indexing capability to pbench. Details:

1. `index-pbench` now calls a module - `index-sar`. 
2. Index name in ElasticSearch for pbench's indices, was modified to accommodate for SAR data mapping.
3. All indexed data goes to pbench.sar-yyyy-mm-dd 
4. sar and pbench runs are connected by the md5 of the pbench run. 

- sar docs in ES have 2 fields from the pbench run:
  1. sar data hierarchy path (as `_metadata.pbench_run_unique_id`) depicting path to sar file in results
  2. pbench-run's md5 (as `_metadata.pbench_run_md5` )

### CODE REVIEW

**Comment by arcolife** [comment]
okay it seems I forgot about the unittests there. Shall see to that..

**Comment by jeremyeder** [comment]
No networking stats?  Well there's a network.json but it doesn't include the kbps or packets per second rx and tx statistics (the most important ones).

**Comment by arcolife** [comment]
Using this, a POST query made to _all indices (for example), in the following form:

```js
{
  "query": {
    "multi_match": {
      "query": "2e2e8933451409eeda920ab6c35f7cde",
      "fields": [
        "_metadata.md5",
        "_metadata.pbench_run_md5"
      ]
    }
  },
  "from": 0,
  "size": 2,
  "fields": [
    "_metadata.file-name",
    "_metadata.nodename"
  ],
  "sort": [],
  "aggs": {}
}
```

..would result in : 

```js
...
    "hits" : [ {
      "_index" : "pbench.2015-10",
      "_type" : "pbench-run",
      "_id" : "2e2e8933451409eeda920ab6c35f7cde",
      "_score" : 3.925263,
      "fields" : {
        "_metadata.file-name" : [ "/var/lib/pbench-agent/benchserver2_sample_pbench_run_2015-10-11.tar.xz" ]
      }
    }, {
      "_index" : "pbench.sar-2015-10-11",
      "_type" : "sar",
      "_id" : "cd703c87c70ea9fe623ce748b4af1ca1",
      "_score" : 2.7668054,
      "fields" : {
        "_metadata.nodename" : [ "benchserver2" ]
      }
    } ]
...
```

**Comment by portante** [comment]
@arcolife, can you rebase this commit?

**Comment by portante** [comment]
And then add unit tests?

**Comment by arcolife** [comment]
rebased. will add the unit tests. Although, is https://travis-ci.org/distributed-system-analysis/pbench/jobs/145559700#L907 this related to this PR? I've faced this before.

**Comment by portante** [comment]
@arcolife, I have seen that before, but I don't remember the root cause.  We'll have to debug it.

**Comment by arcolife** [comment]
update: failure in pbench unit test [ref: https://github.com/distributed-system-analysis/pbench/pull/176#issuecomment-233359641 ] due to modification of index-pbench by this commit and consequently change in expected line numbers under pbench/bin/gold/test-*.txt (viz-a-viz test case output verification texts) ..and so the unittests:_verify_output() (code ref: https://github.com/distributed-system-analysis/pbench/blob/master/server/pbench/bin/unittests#L93 ) declares failed cases.

**Comment by arcolife** [comment]
WIP: unit test rework. reference: https://github.com/distributed-system-analysis/pbench/issues/303
will be adding a proper gold/test-9.txt and uncomment test-9 from bin/unittests

**Comment by arcolife** [comment]
test-9 (for this PR) fails at https://travis-ci.org/distributed-system-analysis/pbench/jobs/150102944#L784
..passes locally on laptop. A difference of time config on machines possibly an issue. Not sure.

**Comment by portante** [comment]
@arcolife, I tried this out this morning, and this is due to your timezone. :(  I think the sadf command needs to be invoked with TZ=UTC, I'll post a PR with a fix to your repo, so that you can pull that in adding it to this PR.

**Comment by portante** [comment]
Okay, your branch need to be rebased on top of upstream master before we can work this through.  Can you do that so I can send you a PR?  @arcolife

**Comment by arcolife** [comment]
@portante thanks for the PR! I've rebased and merged as needed.

**Comment by portante** [comment]
We have to adjust the sar indexing to not use _timestamp field, and to rename the _metadata file to  something else (name TBD).

**Comment by portante** [comment]
Perhaps update sarjitsu to do the same.

**Comment by portante** [comment]
[Merge Archit's sar indexing and grafana work onto UI branch](https://trello.com/c/QhF2Qy2y/4-merge-archit-s-sar-indexing-and-grafana-work-onto-ui-branch)

**Comment by portante** [comment]
[pbench-sar indexing work](https://trello.com/c/OtOjdBi1/2-pbench-sar-indexing-work)

**Comment by portante** [comment]
@arcolife, where are we at with rebasing this work on the `wip-indexing` branch?

**Comment by arcolife** [comment]
@portante I had rebased this against master earlier. And now, I've rebased this against wip-indexing. closing this one and opening a PR against `wip-indexing` instead so @tenstormavi could continue from there. (IIUC, i.e the required workflow)

**Comment by portante** [comment]
Re-opening, gotta save the low PR number!

File: server/pbench/bin/index-pbench [hunk 1]
```diff
@@ -1160,7 +1160,7 @@ def mk_user_specified_metadata(options):
 
 def mk_metadata(fname, tb, mdconf, md5sum):
     """Return a dict with metadata about a tarball"""
-    mtime = datetime.fromtimestamp(os.stat(fname)[stat.ST_MTIME])
+    mtime = datetime.utcfromtimestamp(os.stat(fname)[stat.ST_MTIME])
```
Comment (portante) [inline server/pbench/bin/index-pbench]: Can you make this change a separate PR first?

File: server/pbench/bin/gold/test-7.8.txt [hunk 2]
```diff
@@ -1039,7 +1039,19 @@ len(actions) = 19
         },
         "_type": "pbench-tool-data-iostat"
     }
-]
+]could not convert binary: /var/tmp/pbench-test-server/tmp/uperf__2016-10-06_16:34:03/1-tcp_stream-16384B-32i/sample1/tools-default/flat8_40gb/sar/sar.data
```
Comment (portante) [inline server/pbench/bin/gold/test-7.8.txt]: Let's get this sa binary file issue resolved.  And we have to do it with out carrying around binary, pre-compiled versions of the sadf commands in the tree.

I think we need to tell folks how to name them and place them on a certain path to get that compatibility, but in our tree just use one from the environment.

File: server/pbench/bin/index-pbench (hunk 1 above)
Comment (arcolife) [inline server/pbench/bin/index-pbench]: https://github.com/distributed-system-analysis/pbench/pull/528

File: server/pbench/bin/gold/test-7.8.txt (hunk 2 above)
Comment (arcolife) [inline server/pbench/bin/gold/test-7.8.txt]: so i've removed all sar data from uperf results for now. (old ref: https://github.com/distributed-system-analysis/pbench/pull/523#discussion_r106825374). Do you think we should submit another PR for removing this added sadf compatibility feature from the tree? (and subsequently adding set of instructions to go achieve the same?). 

I'm thinking users would be looking all over the place for older sadf versions (in case they're indexing old results) and presence of / pointer to, these binaries is crucial. Although I guess such cases could be rare.

**Comment by portante** [comment]
Any further updates based on the comments made here so far, @arcolife?

File: server/pbench/bin/detect_nodename [hunk 3]
```diff
@@ -0,0 +1,3 @@
+#!/bin/bash
```
Comment (arcolife) [inline server/pbench/bin/detect_nodename]: So I was sifting through the existing scripts and I'm curious why don't we have this license header in them? case in point, this [index-pbench](https://github.com/distributed-system-analysis/pbench/blob/master/server/pbench/bin/index-pbench#L2) script 
Maybe I still don't fully get the concept of individual tagging of files with licenses. Because I've often come across repos that have a single LICENSE file in project root and that's the end of it..

**Comment by arcolife** [comment]
`extractsa` is now a python script (`extract_sa.py`) that handles various scenarios as per https://github.com/distributed-system-analysis/pbench/pull/176#discussion_r107421658

Usage:
```
 ./extract_sa.py -h
Usage: extract_sa.py [--config <path-to-pbench-server.cfg>] <path-to-sa-binary>

Options:
  -h, --help            show this help message and exit
  -c CFG_NAME, --config=CFG_NAME
                        Specify config file
```

Total incompatibility when a required specific sadf script is not found:
```
Invalid system activity file: /home/arcolife/Documents/sar_samples/test/sa04_65_local
File created by sar/sadc from sysstat version 9.0.4
Current sysstat version cannot read the format of this file (0x1170)

Cannot convert the format of this file

converted binary to current version of sysstat..
None
Invalid system activity file: /home/arcolife/Documents/sar_samples/test/sa04_65_local_conv

Failed to extract converted binary. Checking oscode..
compatible OS version for this binary: 65
Appropriate sadf-<type>-<arch> script necessary to process this file not found!
Please provide a path to folder containing sadf script corresponding to Red Hat OS: 65
Path should be configured under [pbench-sysstat] section in ../lib/config/pbench-server.cfg.example
```

Total incompatibility but sadf script found to deal with this:
```
Invalid system activity file: /home/arcolife/Documents/sar_samples/test/sa04_65_local
File created by sar/sadc from sysstat version 9.0.4
Current sysstat version cannot read the format of this file (0x1170)

Cannot convert the format of this file

converted binary to current version of sysstat..
None
Invalid system activity file: /home/arcolife/Documents/sar_samples/test/sa04_65_local_conv

Failed to extract converted binary. Checking oscode..
compatible OS version for this binary: 65

Nodename: localhost.example.com
XML data saved to: /home/arcolife/Documents/sar_samples/test/sa04_65_gprfc039.xml
```

Partial incompatibility, dealt with conversion:
```
 ./extract_sa.py -c ../lib/config/pbench-server.cfg.example ~/Documents/sar_samples/test/datafile001 
Invalid system activity file: /home/arcolife/Documents/sar_samples/test/datafile001
File created by sar/sadc from sysstat version 10.1.5
Current sysstat version cannot read the format of this file (0x2171)

file_magic: OK
file_header: OK
file_activity: A_CPU A_PCSW A_IRQ A_SWAP A_PAGE A_IO A_MEMORY A_HUGE A_KTABLES A_QUEUE A_DISK A_NET_DEV A_NET_EDEV A_NET_NFS A_NET_NFSD A_NET_SOCK A_NET_IP A_NET_EIP A_NET_ICMP A_NET_EICMP A_NET_TCP A_NET_ETCP A_NET_UDP A_NET_SOCK6 A_NET_IP6 A_NET_EIP6 A_NET_ICMP6 A_NET_EICMP6 A_NET_UDP6 A_PWR_CPUFREQ A_PWR_FAN A_PWR_TEMP A_PWR_IN A_PWR_USB OK
Statistics: .........................................................................................................................
File successfully converted to sysstat format version 11.2.0

converted binary to current version of sysstat..
None

Nodename: benchserver2
XML data saved to: /home/arcolife/Documents/sar_samples/test/datafile001.xml
```
Compatible versions:
```
 ./extract_sa.py -c ../lib/config/pbench-server.cfg.example ~/Documents/sar_samples/test/sa10 

Nodename: localhost.localdomain
XML data saved to: /home/arcolife/Documents/sar_samples/test/sa10.xml
```

next up: need to refactor `index-pbench` and update `sysstat.py` and check unit tests to accommodate this new change.

**Comment by arcolife** [comment]
- [x] update unit tests to accommodate this new change.
  - [ ] https://travis-ci.org/distributed-system-analysis/pbench/jobs/224676630#L876
  - [x] also check if pbench-sysstat is installed (same version) in Travis as on local machine (nope! it's not up to date on Travis, hence the failure in unit test above. Gotta add checks [noted here](https://github.com/arcolife/pbench/blob/sar-index/server/pbench/bin/satools/extract_sa.py#L22))
- [x] update `sysstat.py` for new file magic
- [x] refactor `index-pbench`
- [x] update satools (if possible/needed) with https://github.com/distributed-system-analysis/satools 
  - [ ] [WIP] https://github.com/distributed-system-analysis/satools/pull/1

**Comment by portante** [comment]
Replaced by #645.
